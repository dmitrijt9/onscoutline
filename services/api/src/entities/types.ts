/**
 * Common date format with time part - https://cs.wikipedia.org/wiki/ISO_8601
 * @example 2022-02-24T16:00:00.000Z
 */
export type ISO8601 = string

/**
 * Common date format without time part - https://cs.wikipedia.org/wiki/ISO_8601
 * @example 2022-02-24
 */
export type ISO8601_NoTime = string
