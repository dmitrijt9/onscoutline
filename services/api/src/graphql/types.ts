import { Container } from '../dependency/container/index'
import { ExpressContext } from 'apollo-server-express'

export type GraphqlContext = {
    container: Container
    // user: User
    // loaders: ContextLoaders
    express?: ExpressContext
}
