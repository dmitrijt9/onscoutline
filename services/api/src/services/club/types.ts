export type NewClubRequest = {
    facrId: string
    name: string
    facrUuid: string
}
