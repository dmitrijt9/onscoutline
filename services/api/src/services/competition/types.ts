export type NewCompetitionRequest = {
    regionName: string
    regionId: string
    facrId: string
    name: string
    facrUuid: string
}
