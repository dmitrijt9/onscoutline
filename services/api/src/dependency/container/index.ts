import { Connection, getCustomRepository } from 'typeorm'
import { CompetitionRepository } from '../../repositories/CompetitionRepository'
import { FacrScraper } from '../../services/scrapers/FacrScraper'
import { Scraper } from '../../services/scrapers/Scraper'
import { AppConfig, getAppConfig } from '../config/index'
import { bootstrapDbConnection } from './bootstrap/db-connection'

export const createContainer = async (
    appConfig = getAppConfig(process.env),
): Promise<Container> => {
    const typeormConnection = await bootstrapDbConnection(appConfig)

    const competitionRepository = getCustomRepository(CompetitionRepository)

    const scraper = new Scraper()
    const facrScraper = new FacrScraper(scraper, appConfig, competitionRepository)

    return {
        config: appConfig,
        typeormConnection,
        competitionRepository,
        facrScraper: facrScraper,
    }
}

export interface Container {
    config: AppConfig
    typeormConnection: Connection
    competitionRepository: CompetitionRepository
    facrScraper: FacrScraper
}
