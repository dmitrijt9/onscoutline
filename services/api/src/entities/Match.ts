import { Column, Entity, ManyToOne, PrimaryGeneratedColumn } from 'typeorm'
import { Club } from './Club'
import { CompetitionHasSeason } from './Relations/CompetitionHasSeason'
import { ISO8601 } from './types'

@Entity()
export class Match {
    @PrimaryGeneratedColumn()
    id: number

    @Column('datetime')
    when: ISO8601

    @Column({ nullable: true })
    where?: string

    @ManyToOne(() => Club)
    homeTeam: Club

    @ManyToOne(() => Club)
    awayTeam: Club

    @Column('int')
    scoreHome: number

    @Column('int')
    scoreAway: number

    @ManyToOne(() => CompetitionHasSeason, { nullable: true })
    competitionSeason?: CompetitionHasSeason
}
