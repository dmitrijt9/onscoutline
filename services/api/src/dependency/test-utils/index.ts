import { ClubRepository } from '../../repositories/club/ClubRepository'
import { CompetitionHasSeasonRepository } from '../../repositories/competition/CompetitionHasSeasonRepository'
import { CompetitionRepository } from '../../repositories/competition/CompetitionRepository'
import { MatchRepository } from '../../repositories/match/MatchRepository'
import { PlayerInClubRepository } from '../../repositories/player/PlayerInClubRepository'
import { PlayerInMatchRepository } from '../../repositories/player/PlayerInMatchRepository'
import { PlayerRepository } from '../../repositories/player/PlayerRepository'
import { SeasonRepository } from '../../repositories/season/SeasonRepository'
import { PlayerGameStatisticRepository } from '../../repositories/statistic/PlayerGameStatisticRepository'
import { ExpressApplication } from '../application/ExpressApplication'
import { getAppConfig } from '../config/index'
import { Container, createContainer } from '../container/index'
import { Repository } from 'typeorm'
import supertest from 'supertest'

export type TestingClient = {
    application: ExpressApplication
    request: supertest.SuperTest<supertest.Test>
    container: Container
}

export const createTestingClient = async (container?: Container): Promise<TestingClient> => {
    const baseConfig = getAppConfig(process.env)
    const testContainer = container ?? {
        ...(await createContainer({
            ...baseConfig,
            db: {
                ...baseConfig.db,
                typeorm: {
                    ...baseConfig.db.typeormTest,
                },
            },
        })),
    }

    const application = new ExpressApplication(testContainer)
    const app = await application.start()

    return {
        application,
        request: supertest(app.expressApp),
        container: app.container,
    }
}

export const stopTestApplication = async (application: ExpressApplication) => {
    return application.stop()
}

export const cleanDb = async (client: TestingClient) => {
    const { typeormConnection } = client.container

    const repositories = [
        typeormConnection.getCustomRepository(PlayerInClubRepository),
        typeormConnection.getCustomRepository(PlayerGameStatisticRepository),
        typeormConnection.getCustomRepository(PlayerInMatchRepository),
        typeormConnection.getCustomRepository(MatchRepository),
        typeormConnection.getCustomRepository(CompetitionHasSeasonRepository),
        typeormConnection.getCustomRepository(SeasonRepository),
        typeormConnection.getCustomRepository(CompetitionRepository),
        typeormConnection.getCustomRepository(ClubRepository),
        typeormConnection.getCustomRepository(PlayerRepository),
    ]

    for (const repository of repositories) {
        await (repository as Repository<{}>).delete({})
    }
}
