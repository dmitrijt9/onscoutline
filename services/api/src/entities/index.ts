import { Club } from './Club'
import { Competition } from './Competition'
import { Match } from './Match'
import { Player } from './Player'
import { CompetitionHasSeason } from './Relations/CompetitionHasSeason'
import { PlayerInClub } from './Relations/PlayerInClub'
import { PlayerInMatch } from './Relations/PlayerInMatch'
import { Season } from './Season'

export const entities = [
    Club,
    Competition,
    Player,
    PlayerInClub,
    Match,
    PlayerInMatch,
    Season,
    CompetitionHasSeason,
]
