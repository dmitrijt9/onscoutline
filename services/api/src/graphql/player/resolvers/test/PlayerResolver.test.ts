import faker from '@faker-js/faker'
import { Container } from '../../../../dependency/container/index'
import {
    cleanDb,
    createTestingClient,
    stopTestApplication,
    TestingClient,
} from '../../../../dependency/test-utils/index'
import { Match } from '../../../../entities/Match'
import { PlayerWithMatchInfo } from '../../../../services/match/types'
import { toOnscoutlineDateFormat } from '../../../../utils/conversions'
import { randomTestClub } from '../../../../utils/test-utils/random-club'
import { randomTestPlayer } from '../../../../utils/test-utils/random-player'

describe('Player resolver', () => {
    let testingClient: TestingClient
    let container: Container

    let randomMatch: Match
    let randomPlayersWithmatchInfo: PlayerWithMatchInfo[]

    afterAll(async () => {
        await stopTestApplication(testingClient.application)
    })

    beforeAll(async () => {
        testingClient = await createTestingClient()
        container = testingClient.container

        await cleanDb(testingClient)
        const {
            playerRepository,
            clubRepository,
            playerInClubRepository,
            matchRepository,
            playerService,
        } = container

        const [player1, player2, player3, player4, player5, player6] = await playerRepository.save(
            [1, 2, 3, 4, 5, 6].map((t) => randomTestPlayer()),
        )

        const [randomClub1, randomClub2] = await clubRepository.save(
            [1, 2].map((t) => randomTestClub()),
        )

        await playerInClubRepository.save([
            {
                club: randomClub1,
                player: player1,
                playingFrom: toOnscoutlineDateFormat(faker.date.past(2)),
            },
            {
                club: randomClub1,
                player: player2,
                playingFrom: toOnscoutlineDateFormat(faker.date.past(2)),
            },
            {
                club: randomClub1,
                player: player3,
                playingFrom: toOnscoutlineDateFormat(faker.date.past(2)),
            },
            {
                club: randomClub2,
                player: player4,
                playingFrom: toOnscoutlineDateFormat(faker.date.past(2)),
            },
            {
                club: randomClub2,
                player: player5,
                playingFrom: toOnscoutlineDateFormat(faker.date.past(2)),
            },
            {
                club: randomClub2,
                player: player6,
                playingFrom: toOnscoutlineDateFormat(faker.date.past(2)),
            },
        ])

        randomMatch = await matchRepository.save({
            homeTeam: randomClub1,
            awayTeam: randomClub2,
            scoreAway: 2,
            scoreHome: 2,
            when: '2022-02-02T15:00:00Z',
            facrUuid: 'deiojdi',
        })

        randomPlayersWithmatchInfo = [
            {
                ...player1,
                matchInfo: {
                    fullname: `${player1.surname} ${player1.name}`,
                    goals: [{ minute: 67, type: 'Vlastní' }],
                    isInStartingLineup: true,
                    position: 'Brankář',
                    shirt: 1,
                    side: 'home',
                    substitution: '68',
                    redCardMinute: null,
                    yellowCardMinutes: null,
                },
            },
            {
                ...player2,
                matchInfo: {
                    fullname: `${player2.surname} ${player2.name}`,
                    goals: [{ minute: 2, type: 'Branka' }],
                    isInStartingLineup: true,
                    position: 'Obránce',
                    shirt: 4,
                    side: 'home',
                    substitution: '45',
                    redCardMinute: null,
                    yellowCardMinutes: null,
                },
            },
            {
                ...player3,
                matchInfo: {
                    fullname: `${player3.surname} ${player3.name}`,
                    goals: [{ minute: 46, type: 'Pokutový kop' }],
                    isInStartingLineup: false,
                    position: 'Neuvedeno',
                    shirt: 9,
                    side: 'home',
                    substitution: '45',
                    redCardMinute: null,
                    yellowCardMinutes: [67],
                },
            },
            {
                ...player4,
                matchInfo: {
                    fullname: `${player4.surname} ${player4.name}`,
                    goals: [],
                    isInStartingLineup: true,
                    position: 'Brankář',
                    shirt: 1,
                    side: 'away',
                    substitution: null,
                    redCardMinute: null,
                    yellowCardMinutes: null,
                },
            },
            {
                ...player5,
                matchInfo: {
                    fullname: `${player5.surname} ${player5.name}`,
                    goals: [],
                    isInStartingLineup: true,
                    position: 'Obránce',
                    shirt: 5,
                    side: 'away',
                    substitution: null,
                    redCardMinute: null,
                    yellowCardMinutes: null,
                },
            },
            {
                ...player6,
                matchInfo: {
                    fullname: `${player6.surname} ${player6.name}`,
                    goals: [{ minute: 90, type: 'Branka' }],
                    isInStartingLineup: true,
                    position: 'Útočník',
                    shirt: 10,
                    side: 'away',
                    substitution: null,
                    redCardMinute: null,
                    yellowCardMinutes: null,
                },
            },
        ]

        await playerService.resolvePlayersInMatch(randomPlayersWithmatchInfo, randomMatch)
    })

    it('should return players stats correctly', async () => {
        const playerStatsRequest = `{
          player(id: "6") {
            id
            playerInfo {
              shirt
            }
            stats {
              startingEleven
              scoredGoals
              ownGoals
              penaltyGoals
              playedMatches
              goalsPerGameRatio
              yellowCards
              redCards
              hattricks
              cleanSheets
            }
          }
        }`
        await testingClient.request
            .post(`/graphql`)
            .send({ query: playerStatsRequest })
            .expect(200)
            .then((res) => {
                expect(res.body).toMatchSnapshot()
            })
    })
})
