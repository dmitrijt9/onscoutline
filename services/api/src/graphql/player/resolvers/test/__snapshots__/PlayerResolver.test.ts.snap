// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Player resolver should return players stats correctly 1`] = `
Object {
  "data": Object {
    "player": Object {
      "id": "6",
      "playerInfo": Object {
        "shirt": 10,
      },
      "stats": Object {
        "cleanSheets": 0,
        "goalsPerGameRatio": 1,
        "hattricks": 0,
        "ownGoals": 0,
        "penaltyGoals": 0,
        "playedMatches": 1,
        "redCards": 0,
        "scoredGoals": 1,
        "startingEleven": 1,
        "yellowCards": 0,
      },
    },
  },
}
`;
